File: server/src/main/java/com/tasktracker/service/TaskService.java
package com.tasktracker.service;

import com.tasktracker.dto.TaskDTO;
import com.tasktracker.entity.Task;
import com.tasktracker.entity.Task.Priority;
import com.tasktracker.entity.User;
import com.tasktracker.repository.TaskRepository;
import com.tasktracker.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * TaskService - Handles all task-related business logic
 */
@Service
public class TaskService {

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private UserRepository userRepository;

    /**
     * Get all tasks for a user
     */
    public List<TaskDTO> getAllTasks(Long userId) {
        return taskRepository.findByUserId(userId)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    /**
     * Create a new task
     */
    public TaskDTO createTask(TaskDTO taskDTO, Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Task task = Task.builder()
                .title(taskDTO.getTitle())
                .description(taskDTO.getDescription())
                .priority(taskDTO.getPriority())
                .dueDate(taskDTO.getDueDate())
                .completed(false)
                .user(user)
                .build();

        return toDTO(taskRepository.save(task));
    }

    /**
     * Update an existing task
     */
    public TaskDTO updateTask(Long taskId, TaskDTO taskDTO, Long userId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new RuntimeException("Task not found"));

        // Verify ownership
        if (!task.getUser().getId().equals(userId)) {
            throw new RuntimeException("Unauthorized: You can only edit your own tasks");
        }

        // Update fields
        task.setTitle(taskDTO.getTitle());
        task.setDescription(taskDTO.getDescription());
        task.setPriority(taskDTO.getPriority());
        task.setDueDate(taskDTO.getDueDate());
        task.setCompleted(taskDTO.isCompleted());

        return toDTO(taskRepository.save(task));
    }

    /**
     * Delete a task
     */
    public void deleteTask(Long taskId, Long userId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new RuntimeException("Task not found"));

        // Verify ownership
        if (!task.getUser().getId().equals(userId)) {
            throw new RuntimeException("Unauthorized: You can only delete your own tasks");
        }

        taskRepository.deleteById(taskId);
    }

    /**
     * Search tasks by title keyword
     */
    public List<TaskDTO> searchTasks(Long userId, String keyword) {
        return taskRepository.searchByTitle(userId, keyword)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    /**
     * Filter tasks by completion status and/or priority
     */
    public List<TaskDTO> filterTasks(Long userId, Boolean completed, String priority) {
        List<Task> tasks;

        if (completed != null && priority != null) {
            Priority p = Priority.valueOf(priority.toUpperCase());
            tasks = taskRepository.findByUserIdAndCompletedAndPriority(userId, completed, p);
        } else if (completed != null) {
            tasks = taskRepository.findByUserIdAndCompleted(userId, completed);
        } else if (priority != null) {
            Priority p = Priority.valueOf(priority.toUpperCase());
            tasks = taskRepository.findByUserIdAndPriority(userId, p);
        } else {
            tasks = taskRepository.findByUserId(userId);
        }

        return tasks.stream().map(this::toDTO).collect(Collectors.toList());
    }

    /**
     * Get dashboard statistics for a user
     */
    public Map<String, Object> getDashboardStats(Long userId) {
        long total = taskRepository.countByUserId(userId);
        long completed = taskRepository.countByUserIdAndCompleted(userId, true);
        long pending = total - completed;
        double progress = total > 0 ? (double) completed / total * 100 : 0;

        Map<String, Object> stats = new HashMap<>();
        stats.put("total", total);
        stats.put("completed", completed);
        stats.put("pending", pending);
        stats.put("progress", Math.round(progress));

        return stats;
    }

    /**
     * Convert Task entity to TaskDTO
     */
    private TaskDTO toDTO(Task task) {
        TaskDTO dto = new TaskDTO();
        dto.setId(task.getId());
        dto.setTitle(task.getTitle());
        dto.setDescription(task.getDescription());
        dto.setPriority(task.getPriority());
        dto.setDueDate(task.getDueDate());
        dto.setCompleted(task.isCompleted());
        dto.setCreatedAt(task.getCreatedAt());
        dto.setUserId(task.getUser().getId());
        return dto;
    }
}

File: client/src/services/taskService.js
// ============================================
// taskService.js - Handles all task-related API calls
// ============================================

import axios from 'axios';

const BASE_URL = 'http://localhost:8080/api/tasks';

/**
 * Create axios instance with JWT auth header
 */
const getAuthHeaders = () => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
  };
};

/**
 * Fetch all tasks for the logged-in user
 */
export const getAllTasks = async () => {
  const response = await axios.get(BASE_URL, getAuthHeaders());
  return response.data;
};

/**
 * Create a new task
 * @param {Object} taskData - { title, description, priority, dueDate }
 */
export const createTask = async (taskData) => {
  const response = await axios.post(BASE_URL, taskData, getAuthHeaders());
  return response.data;
};

/**
 * Update an existing task by ID
 * @param {number} id - Task ID
 * @param {Object} taskData - Updated task fields
 */
export const updateTask = async (id, taskData) => {
  const response = await axios.put(`${BASE_URL}/${id}`, taskData, getAuthHeaders());
  return response.data;
};

/**
 * Delete a task by ID
 * @param {number} id - Task ID
 */
export const deleteTask = async (id) => {
  const response = await axios.delete(`${BASE_URL}/${id}`, getAuthHeaders());
  return response.data;
};

/**
 * Search tasks by keyword in title
 * @param {string} keyword - Search term
 */
export const searchTasks = async (keyword) => {
  const response = await axios.get(`${BASE_URL}/search?keyword=${keyword}`, getAuthHeaders());
  return response.data;
};

/**
 * Filter tasks by completion status and/or priority
 * @param {boolean|null} completed - Filter by status
 * @param {string|null} priority   - Filter by priority (HIGH, MEDIUM, LOW)
 */
export const filterTasks = async (completed, priority) => {
  let url = `${BASE_URL}/filter?`;
  if (completed !== null && completed !== undefined) url += `completed=${completed}&`;
  if (priority) url += `priority=${priority}`;

  const response = await axios.get(url, getAuthHeaders());
  return response.data;
};

/**
 * Fetch dashboard statistics (total, completed, pending, progress)
 */
export const getDashboardStats = async () => {
  const response = await axios.get(`${BASE_URL}/stats`, getAuthHeaders());
  return response.data;
};
